### DISCUSSION TITLE
classes/class_fileaccess

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#200] [User-contributed notes]
# classes/class_fileaccess

Inherits: RefCounted< Object Provides methods for file reading and writing operations. Description: This class can be used to permanently store data in the user device's file system and to read fro...

https://docs.godotengine.org/en/latest/classes/class_fileaccess.html

**Comment by aaron-bigadventurecrew** [comment]
> void store_line(line: String) 
> Appends line to the file followed by a line return character (\n), encoding the text as UTF-8.

The use of the word *Appends* here is not accurate and the intended functionality of only appending a line can only be achieved by adding **seek_end() _before_ store_line()**, i.e.:

```
var file = FileAccess.open("user://sample_file.extension", FileAccess.READ_WRITE)
file.seek_end()
file.store_line("Entered Strings or String Variables here")
```

I have replicated this behavior across FileAccess.WRITE and FileAccess.WRITE_READ as well and as far back as 4.3-beta3 up until the current 4.4-dev3. What's interesting is that without seek_end() it does end with a \n new line, but will keep overwriting the file from the beginning instead of appending new lines.

Entering this feedback as a comment as this is likely a wording issue of how store_line() is defined here in the docs which isn't aligning with what the engine is actually doing. If the intention of store_line() is what I describe when adding seek_end() then I can raise a GitHub issue with a repro project upon request.

Edit 1: I raised a GitHub issue on this with a MRP available for issue replication
[https://github.com/godotengine/godot/issues/98243](https://github.com/godotengine/godot/issues/98243)

Edit 2: A PR has been made by the team to update `store_line(`)'s description
[https://github.com/godotengine/godot/pull/98238](https://github.com/godotengine/godot/pull/98238)
This was previously noted in a V3 PR as well
[https://github.com/godotengine/godot/pull/42442](https://github.com/godotengine/godot/pull/42442)
Will update once more once I see the changes reflected here on class_fileaccess.html

Edit 3: PR was merged by the team and new wording for `store_line()` will be available ~19 Oct 2024. For anyone needing this functionality of only appending to a file with **FileAccess**, you can still use the Godot 4 code above as the need for `seek_end()` is necessary and not indicative of an engine issue in and of itself now that `store_line()`'s definition has been updated.

Thanks so much Godot team!

**Comment by azuriancomet** [comment]
`store_pascal_string()` and by extension `get_pascal_string()` use `uint32_t` to store and retrieve the length of the string, as per: https://github.com/godotengine/godot/blob/master/core/io/file_access.cpp#L714

The first `4 bytes` represent the length of the encoded string , which means you can store strings between 1_073_741_823 and 4_294_967_295 `chars`.

You can test this using the following code:
```gdscript
func _ready() -> void:
	var string := "A bad workman always blames his tools."
	var file := FileAccess.open("res://file.ext", FileAccess.WRITE)
	file.store_pascal_string(string)
	file.close()
	
	var _file := FileAccess.open("res://file.ext", FileAccess.READ)
	var stored_length := _file.get_32() # Get the first 4 bytes
	
	print("String length: %s\nStored length: %s" % [string.length(), stored_length])
```
Beware that `String.length()` returns the number of `chars` in the string, which will be equivalent to the stored length only if the string contains ASCII characters exclusively (since they are the only Unicode block encoded using `1 byte`).

**Comment by MoltenCoffee** [comment]
`store_var()`, although it indeed uses the same functionality as [`@GlobalScope.var_to_bytes()`](https://docs.godotengine.org/en/stable/classes/class_%40globalscope.html#class-globalscope-method-var-to-bytes), does _also_ stores the number of bytes used for storing the variant (used to set a buffer size upon reading), unlike what one might expect by reading [the docs on the serialization API](https://docs.godotengine.org/en/stable/tutorials/io/binary_serialization_api.html).

To illustrate, for an int, `var_to_bytes()` returns 8 bytes, 2 bytes for the type, 2 for flags, 4 for the data

```gdscript
print(var_to_bytes(22))
# [2, 0, 0, 0, 22, 0, 0, 0]
```

But `FileAccess.store_var()` will write 12, with 4 prepended to store the size of the buffer needed for reading the variant:

```gdscript
var fa := FileAccess.open("user://test.bin", FileAccess.WRITE_READ)
fa.store_var(22)
fa.seek(0) # Back to beginning of file
var buffer := fa.get_buffer(fa.get_length())
print(buffer) # Print contents of file as bytes
# [8, 0, 0, 0, 2, 0, 0, 0, 22, 0, 0, 0]

# The following will _not_ work:
print(bytes_to_var(buffer))
# <null>
```

**Comment by Samvkes** [comment]
I think the default "writing" behaviour here (manually moving a pointer from the front of the file, then inserting text) is different from most other file API's I've seen. I think the documentation and the method descriptions should better reflect that.

**Reply by Samvkes** [reply]
(To clarify, I think most other api's focus on either "writing" ie clear the file and add new text, or "appending", ie add text to end of what was already in the file)
